[JSC] Allow CodeBlock hash computation from concurrent compiler thread

Constellation opened the pull request:
#### 4f51c8d055c64c0c4c0343a07d08d4cacf322ea4
<pre>
[JSC] Allow CodeBlock hash computation from concurrent compiler thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=293146">https://bugs.webkit.org/show_bug.cgi?id=293146</a>
<a href="https://rdar.apple.com/problem/151486734">rdar://problem/151486734</a>

Reviewed by Yijia Huang.

CodeBlockHash computation does not rely on any main thread concept. It
can run in the concurrent compiler thread. Thus, we do not need to
prohibit it.
The difficulty is that WebCore main thread could replace the
SourceProvider&apos;s underlying buffer when NetworkProcess cached the
result. It is fine if we are reading source data from the main thread,
but it is not fine when we are reading it from the concurrent compiler
thread since it can be possible that we replace and discard old source
data while the compiler thread is reading it. To avoid that, we already
have a mechanism &quot;SourceProviderBufferGuard&quot;. This was used only for
Wasm before, but this patch extends it to support normal JS source code
so that we can keep the data while reading it from the compiler thread.

* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::inferredNameWithHash const):
(JSC::CodeBlock::hash const):
(JSC::CodeBlock::finishCreation):
(JSC::CodeBlock::isSafeToComputeHash const): Deleted.
(JSC::CodeBlock::hashAsStringIfPossible const): Deleted.
* Source/JavaScriptCore/bytecode/CodeBlock.h:
* Source/JavaScriptCore/bytecode/CodeBlockHash.cpp:
(JSC::CodeBlockHash::CodeBlockHash):
* Source/JavaScriptCore/bytecode/CodeBlockHash.h:
(JSC::CodeBlockHash::operator bool const):
(WTF::StringTypeAdapter&lt;JSC::CodeBlockHash&gt;::StringTypeAdapter):
(WTF::StringTypeAdapter&lt;JSC::CodeBlockHash&gt;::length const):
(WTF::StringTypeAdapter&lt;JSC::CodeBlockHash&gt;::is8Bit const):
(WTF::StringTypeAdapter&lt;JSC::CodeBlockHash&gt;::writeTo const):
(JSC::CodeBlockHash::operator! const): Deleted.
* Source/JavaScriptCore/bytecode/InlineCallFrame.cpp:
(JSC::InlineCallFrame::dumpBriefFunctionInformation const):
(JSC::InlineCallFrame::hashAsStringIfPossible const): Deleted.
* Source/JavaScriptCore/bytecode/InlineCallFrame.h:
* Source/JavaScriptCore/ftl/FTLJITFinalizer.cpp:
(JSC::FTL::JITFinalizer::finalize):
* Source/JavaScriptCore/parser/SourceProvider.cpp:
(JSC::SourceProvider::lockUnderlyingBuffer):
(JSC::SourceProvider::unlockUnderlyingBuffer):
* Source/JavaScriptCore/parser/SourceProvider.h:
(JSC::SourceProviderBufferGuard::SourceProviderBufferGuard):
(JSC::SourceProviderBufferGuard::~SourceProviderBufferGuard):
(JSC::BaseWebAssemblySourceProvider::lockUnderlyingBuffer): Deleted.
(JSC::BaseWebAssemblySourceProvider::unlockUnderlyingBuffer): Deleted.
(JSC::WebAssemblySourceProviderBufferGuard::WebAssemblySourceProviderBufferGuard): Deleted.
(JSC::WebAssemblySourceProviderBufferGuard::~WebAssemblySourceProviderBufferGuard): Deleted.
* Source/JavaScriptCore/runtime/FunctionExecutableDump.cpp:
(JSC::FunctionExecutableDump::dump const):
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/tools/FunctionAllowlist.cpp:
(JSC::FunctionAllowlist::contains const):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyHelpers.h:
(JSC::getWasmBufferFromValue):
(JSC::createSourceBufferFromValue):
* Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
(WebCore::CachedScriptSourceProvider::create): Deleted.
(WebCore::CachedScriptSourceProvider::~CachedScriptSourceProvider): Deleted.
(WebCore::CachedScriptSourceProvider::CachedScriptSourceProvider): Deleted.
(WebCore::CachedScriptSourceProvider::m_cachedScript): Deleted.
* Source/WebCore/bindings/js/ScriptBufferSourceProvider.h:
* Tools/Scripts/compare-optimization-tracing:
(run_jsc_and_capture):

Canonical link: <a href="https://commits.webkit.org/295048@main">https://commits.webkit.org/295048@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/256e649a25ea9bfae5ca3422a1694231d385a6ab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103918 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23621 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13942 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109111 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/54579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105958 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23976 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32166 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78941 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/54579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106924 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18616 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93742 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/59269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11794 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53946 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/96593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11852 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111498 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/102529 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31074 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/22905 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87951 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31437 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89943 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87606 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32495 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10230 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/25440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16869 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31003 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/36313 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/126162 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30796 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34912 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34133 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/32357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 reviewed (approved):
r=me if EWS is happy.

MenloDorian commented on Source/JavaScriptCore/parser/SourceProvider.h:
```diff
@@ -206,23 +206,25 @@ class UnlinkedFunctionCodeBlock;
     };
 
     // RAII class for managing a Wasm source provider's underlying buffer.
```
Can you fix this comment?  It's no longer just Wasm source.

Constellation replied:
Fixed!

Constellation committed:
[JSC] Allow CodeBlock hash computation from concurrent compiler thread
https://bugs.webkit.org/show_bug.cgi?id=293146
rdar://problem/151486734

Reviewed by Yijia Huang.

CodeBlockHash computation does not rely on any main thread concept. It
can run in the concurrent compiler thread. Thus, we do not need to
prohibit it.
The difficulty is that WebCore main thread could replace the
SourceProvider's underlying buffer when NetworkProcess cached the
result. It is fine if we are reading source data from the main thread,
but it is not fine when we are reading it from the concurrent compiler
thread since it can be possible that we replace and discard old source
data while the compiler thread is reading it. To avoid that, we already
have a mechanism "SourceProviderBufferGuard". This was used only for
Wasm before, but this patch extends it to support normal JS source code
so that we can keep the data while reading it from the compiler thread.

* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::inferredNameWithHash const):
(JSC::CodeBlock::hash const):
(JSC::CodeBlock::finishCreation):
(JSC::CodeBlock::isSafeToComputeHash const): Deleted.
(JSC::CodeBlock::hashAsStringIfPossible const): Deleted.
* Source/JavaScriptCore/bytecode/CodeBlock.h:
* Source/JavaScriptCore/bytecode/CodeBlockHash.cpp:
(JSC::CodeBlockHash::CodeBlockHash):
* Source/JavaScriptCore/bytecode/CodeBlockHash.h:
(JSC::CodeBlockHash::operator bool const):
(WTF::StringTypeAdapter<JSC::CodeBlockHash>::StringTypeAdapter):
(WTF::StringTypeAdapter<JSC::CodeBlockHash>::length const):
(WTF::StringTypeAdapter<JSC::CodeBlockHash>::is8Bit const):
(WTF::StringTypeAdapter<JSC::CodeBlockHash>::writeTo const):
(JSC::CodeBlockHash::operator! const): Deleted.
* Source/JavaScriptCore/bytecode/InlineCallFrame.cpp:
(JSC::InlineCallFrame::dumpBriefFunctionInformation const):
(JSC::InlineCallFrame::hashAsStringIfPossible const): Deleted.
* Source/JavaScriptCore/bytecode/InlineCallFrame.h:
* Source/JavaScriptCore/ftl/FTLJITFinalizer.cpp:
(JSC::FTL::JITFinalizer::finalize):
* Source/JavaScriptCore/parser/SourceProvider.cpp:
(JSC::SourceProvider::lockUnderlyingBuffer):
(JSC::SourceProvider::unlockUnderlyingBuffer):
* Source/JavaScriptCore/parser/SourceProvider.h:
(JSC::SourceProviderBufferGuard::SourceProviderBufferGuard):
(JSC::SourceProviderBufferGuard::~SourceProviderBufferGuard):
(JSC::BaseWebAssemblySourceProvider::lockUnderlyingBuffer): Deleted.
(JSC::BaseWebAssemblySourceProvider::unlockUnderlyingBuffer): Deleted.
(JSC::WebAssemblySourceProviderBufferGuard::WebAssemblySourceProviderBufferGuard): Deleted.
(JSC::WebAssemblySourceProviderBufferGuard::~WebAssemblySourceProviderBufferGuard): Deleted.
* Source/JavaScriptCore/runtime/FunctionExecutableDump.cpp:
(JSC::FunctionExecutableDump::dump const):
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/tools/FunctionAllowlist.cpp:
(JSC::FunctionAllowlist::contains const):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyHelpers.h:
(JSC::getWasmBufferFromValue):
(JSC::createSourceBufferFromValue):
* Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
(WebCore::CachedScriptSourceProvider::create): Deleted.
(WebCore::CachedScriptSourceProvider::~CachedScriptSourceProvider): Deleted.
(WebCore::CachedScriptSourceProvider::CachedScriptSourceProvider): Deleted.
(WebCore::CachedScriptSourceProvider::m_cachedScript): Deleted.
* Source/WebCore/bindings/js/ScriptBufferSourceProvider.h:
* Tools/Scripts/compare-optimization-tracing:
(run_jsc_and_capture):

Canonical link: https://commits.webkit.org/295048@main
